src/i_flmusic.c

```diff
@@ -0,0 +1,288 @@
+//
+// Copyright(C) 2022 Roman Fomin
+// Copyright(C) 2023 Michael Day
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//      FluidSynth backend
+
+#include "config.h"
+
+#if HAVE_FLUIDSYNTH
+
+#include "fluidsynth.h"
+
+#if (FLUIDSYNTH_VERSION_MAJOR < 2 ||                                           \
+     (FLUIDSYNTH_VERSION_MAJOR == 2 && FLUIDSYNTH_VERSION_MINOR < 2))
+
+typedef int fluid_int_t;
+typedef long fluid_long_long_t;
+
+#else
+
+typedef fluid_long_long_t fluid_int_t;
+
+#endif
+
+#include "SDL_mixer.h"
+
+#include "doomtype.h"
+#include "i_system.h"
+#include "i_sound.h"
+#include "m_misc.h"
+#include "memio.h"
+#include "mus2mid.h"
+
+char *fsynth_sf_path = "";
+int fsynth_chorus_active = 1;
+float fsynth_chorus_depth = 5.0;
+float fsynth_chorus_level = 0.35;
+int fsynth_chorus_nr = 3;
+float fsynth_chorus_speed = 0.3;
+char *fsynth_midibankselect = "gs";
+int fsynth_polyphony = 256;
+int fsynth_reverb_active = 1;
+float fsynth_reverb_damp = 0.4;
+float fsynth_reverb_level = 0.15;
+float fsynth_reverb_roomsize = 0.6;
+float fsynth_reverb_width = 4.0;
+
+static fluid_synth_t *synth = NULL;
+static fluid_settings_t *settings = NULL;
+static fluid_player_t *player = NULL;
+
+static void FL_Mix_Callback(void *udata, Uint8 *stream, int len)
+{
+    int result;
+
+    result = fluid_synth_write_s16(synth, len / 4, stream, 0, 2, stream, 1, 2);
+
+    if (result != FLUID_OK)
+    {
+        fprintf(stderr,
+                "FL_Mix_Callback: Error generating FluidSynth audio.\n");
+    }
+}
+
+static boolean I_FL_InitMusic(void)
+{
+    int sf_id;
+
+    if (strlen(fsynth_sf_path) == 0)
+    {
+        fprintf(stderr,
+                "I_FL_InitMusic: No FluidSynth soundfont file specified.\n");
+        return false;
+    }
+
+    // quiet FluidSynth
+    fluid_set_log_function(FLUID_ERR, NULL, NULL);
```

Comment:
Should there be some kind of option to log whatever fluidsynth is telling us for debugging purposes?